File: server/reset-password.mjs
import { openDatabase } from './database.mjs';
import { hashPassword, normalizeEmail, validPassword } from './security.mjs';
import { resolve } from 'node:path';
import { existsSync } from 'node:fs';
import { createInterface } from 'node:readline/promises';
import { Writable } from 'node:stream';
if(!process.stdin.isTTY)throw new Error('Bitte in einem interaktiven Terminal ausführen.');
const database=resolve(process.env.DATA_DIR||'data','returns.sqlite');
if(!existsSync(database))throw new Error('Datenbank nicht gefunden.');
let muted=false;
const output=new Writable({write(chunk,encoding,done){if(!muted)process.stdout.write(chunk);done();}});
const rl=createInterface({input:process.stdin,output,terminal:true});
let email,password;
try{
 email=normalizeEmail(await rl.question('E-Mail des bestehenden Kontos: '));
 process.stdout.write('Neues Passwort (Eingabe verborgen): ');muted=true;password=await rl.question('');muted=false;process.stdout.write('\n');
}finally{rl.close();}
if(!validPassword(password))throw new Error('Passwort muss 12–128 Zeichen enthalten.');
const db=openDatabase(database);
try{
 const user=db.prepare('SELECT id FROM users WHERE email=?').get(email);
 if(!user)throw new Error('Dieses Konto existiert nicht.');
 const encoded=await hashPassword(password);
 db.exec('BEGIN IMMEDIATE');
 try{db.prepare('UPDATE users SET password_hash=? WHERE id=?').run(encoded,user.id);db.prepare('DELETE FROM sessions WHERE user_id=?').run(user.id);db.exec('COMMIT');}catch(e){db.exec('ROLLBACK');throw e;}
 console.log('Passwort geändert. Bestehende Sitzungen wurden abgemeldet.');
}finally{db.close();}

File: server/reset_password.py
import getpass
import os
from pathlib import Path
import sys
from dotenv import load_dotenv
from .database import connect, transaction
from .security import hash_password, normalize_email, valid_password


def main():
    if not sys.stdin.isatty():
        raise SystemExit('Bitte in einem interaktiven Terminal ausführen.')
    load_dotenv()
    database = Path(os.environ.get('DATA_DIR', 'data')) / 'returns.sqlite'
    if not database.is_file():
        raise SystemExit('Datenbank nicht gefunden.')
    email = normalize_email(input('E-Mail des bestehenden Kontos: '))
    password = getpass.getpass('Neues Passwort (Eingabe verborgen): ')
    if not valid_password(password):
        raise SystemExit('Passwort muss 12–128 Zeichen enthalten.')
    db = connect(database)
    try:
        user = db.execute('SELECT id FROM users WHERE email=?', (email,)).fetchone()
        if not user:
            raise SystemExit('Dieses Konto existiert nicht.')
        encoded = hash_password(password)
        with transaction(db):
            db.execute('UPDATE users SET password_hash=? WHERE id=?', (encoded, user['id']))
            db.execute('DELETE FROM sessions WHERE user_id=?', (user['id'],))
        print('Passwort geändert. Bestehende Sitzungen wurden abgemeldet.')
    finally:
        db.close()


if __name__ == '__main__':
    main()
